src/Elastic.Apm/Helpers/SystemInfoHelper.cs

```diff
@@ -0,0 +1,106 @@
+using System;
+using System.IO;
+using System.Text.RegularExpressions;
+using Elastic.Apm.Api;
+using Elastic.Apm.Logging;
+
+namespace Elastic.Apm.Helpers
+{
+	internal class SystemInfoHelper
+	{
+		private const string ContainerUidRegex = "^[0-9a-fA-F]{64}$";
+
+		private const string PodRegex = "(?:^/kubepods/[^/]+/pod([^/]+)$)|"
+			+ "(?:^/kubepods\\.slice/kubepods-[^/]+\\.slice/kubepods-[^/]+-pod([^/]+)\\.slice$)";
+
+		private static Api.System ParseContainerId(string line)
+		{
+			//Copied from the Java agent, and C#-ified
+
+			string kubernetesPodUid = null;
+
+			var fields = line.Split(':');
+			if (fields.Length != 3) return null;
+
+			var cGroupPath = fields[2];
+
+
+			var idPart = Path.GetFileName(cGroupPath);
+
+			if (string.IsNullOrWhiteSpace(idPart)) return null;
+
+			// Legacy, e.g.: /system.slice/docker-<CID>.scope
+			if (idPart.EndsWith(".scope"))
+				idPart = idPart.Substring(0, idPart.Length - ".scope".Length).Substring(idPart.IndexOf("-", StringComparison.Ordinal) + 1);
+
+			// Looking for kubernetes info
+			var dirPathPart = Directory.GetParent(cGroupPath); // Path.get (cGroupPath);  cGroupPath Path.GetPathRoot(cGroupPath);
```

Comment:
I wouldn't recommend using file system API on strings that are not related to file system. For example on Windows for `cGroupPath` with value `/kubepods/besteffort/pode9b90526-f47d-11e8-b2a5-080027b9f4fb/15aa6e53-b09a-40c7-8558-c6c31e36c88a`, `Directory.GetParent(cGroupPath)` is `C:\kubepods\besteffort\pode9b90526-f47d-11e8-b2a5-080027b9f4fb` and of course the whole test fails.

Reply:
`Path.GetDirectoryName(cGroupPath);` would probably a better choice here - we use `Path` everywhere, so with this we rely on that for parsing. What do you think? Tested it on Windows and that works. 

Or by saying 
> I wouldn't recommend using file system API on strings 
you also mean `Path` and not only `Directory`?

Reply:
I would not use file system API here at all because relation to file system is purely coincidental - just because `/` is used in the string. My concern is that file system API is much heavier than just parsing strings and I would avoid bringing heavier API than the necessary minimum. Like I mentioned above I would just implement it by searching backwards for the last `/`. But it's up to you to decide - I'm fine with your proposal if it passes all the tests.

Reply:
I see; well that'd be a bigger refactoring. Also, this is stolen from Java, and they also do it this way. I think it's fine, I'd also prefer using this API than reimplementing it and do those things manually. Let's roll it this way for now 🚀 

cc @eyalkoren @felixbarny

Reply:
@gregkalapos I just ran into an issue with my unit tests failing in Windows because of using the FileSystem APIs. So, while I maintain the usage of those APIs for actual reading the file content (which actually requires real file system operations), I am removing them from the ID parsing part, which deals with stuff that only look like paths, but probably aren't...

Reply:
Thanks for the heads up! Are you working on docker windows support, or is it some other id?

Reply:
Since this is about container IDs, it is a Linux thing anyway. I changed that only to enable building our agent on Windows.

Reply:
Och I see... we have runtime checks and in our case this part only runs on Linux - both the prod code and also tests covering this.